ktlint-core/src/main/kotlin/com/pinterest/ktlint/core/internal/SupressedRegionLocator.kt

```diff
@@ -73,32 +76,19 @@ private fun collect(
             val text = node.getText()
             if (text.startsWith("//")) {
                 val commentText = text.removePrefix("//").trim()
-                parseHintArgs(commentText, "ktlint-disable")?.let { args ->
-                    val lineStart = (
-                        node.prevLeaf { it is PsiWhiteSpace && it.textContains('\n') } as
-                            PsiWhiteSpace?
-                        )?.let { it.node.startOffset + it.text.lastIndexOf('\n') + 1 } ?: 0
-                    result.add(SuppressionHint(IntRange(lineStart, node.startOffset), HashSet(args)))
+                if (node.isDisabledOnPackageStatement()) {
```

Comment:
KtLint readme [on this case](https://github.com/pinterest/ktlint#how-do-i-suppress-an-error-for-a-lineblockfile) is not very clear.
```
import package.* // ktlint-disable
```
Should disable any rule check on this particular line.

It doesn't make sense to disable all the rules for the whole file - better to either not pass it to KtLint or exclude it via GLOB pattern. To disable all the rules for the region, better to use:
```kotlin
/* ktlint-disable */
<some code>
/* ktlint-enable */
```

So, for the case `<some code> // ktlint-disable` please use `createLineDisableSupressionHint` and pass `emptySet()` as `disabledRules`.

Reply:
I really like puzzles. But this review remark really confused me at first. I have the feeling that we were discussing different, albeit related, topics. Let's try to unravel this.

> KtLint readme [on this case](https://github.com/pinterest/ktlint#how-do-i-suppress-an-error-for-a-lineblockfile) is not very clear.
> 
> ```
> import package.* // ktlint-disable
> ```
> 
> Should disable any rule check on this particular line.
This code example is confusing for different reasons:
* There is no apparent reason why this statement should be excluded by ktlint as there is no rule violated.
* The word `package` after `import` took precedence for me. So I read this statement as being the package statement:
```
package com.example.mypackage // ktlint-disable
```
I have modified this.

> It doesn't make sense to disable all the rules for the whole file - better to either not pass it to KtLint or exclude it via GLOB pattern. 
It is only possible to follow this approach when you know which files need to be excluded from processing by ktlint. In bigger code bases with which you are not familiar you may just not know which files would need to be excluded. When you are able to specify it explicitly in the file itself, you can also document the reasoning for this. This approach allows to lint/format the entire code base without having to know the files which need to be excluded.

> So, for the case `<some code> // ktlint-disable` please use `createLineDisableSupressionHint` and pass `emptySet()` as `disabledRules`.
I have removed the code for the `ktlint-disable` directive on the *package* statement as I no longer needed it after fixing some other bugs.

I found out that certain bugs are caused by the fact that some rules perform a kind of initialization as soon as the rootNode is visited. If for some reasons, for example disabling a rule, the rootNode is not visited then the rule would not have been initialized correctly. I have split initialization and visiting.

Next problem was that the IndentationRule modifies the ASTNode hierarchy by inserting additional node. Disabling the indentation rule was not always recognized by the SupressedRegionLocator. I have fixed and simplified this locator.

The changes can best be reviewed per commit.